### PR TITLE
OSCAR integration

### PR DESCRIPTION
**Pull request by catttam**

### CODE REVIEW

**Comment by IgnacioHeredia** [comment]
Hi @catttam ,

just did a fresh install of the `oscar` branch.
Trying to do `from oscar_python.storage import Storage` raises this error:

<details>
  <summary>error</summary>
 
  ```bash
>>> from oscar_python.storage import Storage
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/iheredia/anaconda3/lib/python3.8/site-packages/oscar_python/storage.py", line 15, in <module>
    from oscar_python._providers._minio import Minio
  File "/home/iheredia/anaconda3/lib/python3.8/site-packages/oscar_python/_providers/_minio.py", line 15, in <module>
    import boto3
  File "/home/iheredia/anaconda3/lib/python3.8/site-packages/boto3/__init__.py", line 17, in <module>
    from boto3.session import Session
  File "/home/iheredia/anaconda3/lib/python3.8/site-packages/boto3/session.py", line 17, in <module>
    import botocore.session
  File "/home/iheredia/anaconda3/lib/python3.8/site-packages/botocore/session.py", line 26, in <module>
    import botocore.client
  File "/home/iheredia/anaconda3/lib/python3.8/site-packages/botocore/client.py", line 15, in <module>
    from botocore import waiter, xform_name
  File "/home/iheredia/anaconda3/lib/python3.8/site-packages/botocore/waiter.py", line 18, in <module>
    from botocore.docs.docstring import WaiterDocstring
  File "/home/iheredia/anaconda3/lib/python3.8/site-packages/botocore/docs/__init__.py", line 15, in <module>
    from botocore.docs.service import ServiceDocumenter
  File "/home/iheredia/anaconda3/lib/python3.8/site-packages/botocore/docs/service.py", line 14, in <module>
    from botocore.docs.client import (
  File "/home/iheredia/anaconda3/lib/python3.8/site-packages/botocore/docs/client.py", line 18, in <module>
    from botocore.docs.example import ResponseExampleDocumenter
  File "/home/iheredia/anaconda3/lib/python3.8/site-packages/botocore/docs/example.py", line 13, in <module>
    from botocore.docs.shape import ShapeDocumenter
  File "/home/iheredia/anaconda3/lib/python3.8/site-packages/botocore/docs/shape.py", line 19, in <module>
    from botocore.utils import is_json_value_header
  File "/home/iheredia/anaconda3/lib/python3.8/site-packages/botocore/utils.py", line 37, in <module>
    import botocore.httpsession
  File "/home/iheredia/anaconda3/lib/python3.8/site-packages/botocore/httpsession.py", line 45, in <module>
    from urllib3.contrib.pyopenssl import (
  File "/home/iheredia/anaconda3/lib/python3.8/site-packages/urllib3/contrib/pyopenssl.py", line 50, in <module>
    import OpenSSL.SSL
  File "/home/iheredia/anaconda3/lib/python3.8/site-packages/OpenSSL/__init__.py", line 8, in <module>
    from OpenSSL import crypto, SSL
  File "/home/iheredia/anaconda3/lib/python3.8/site-packages/OpenSSL/crypto.py", line 1556, in <module>
    class X509StoreFlags(object):
  File "/home/iheredia/anaconda3/lib/python3.8/site-packages/OpenSSL/crypto.py", line 1577, in X509StoreFlags
    CB_ISSUER_CHECK = _lib.X509_V_FLAG_CB_ISSUER_CHECK
AttributeError: module 'lib' has no attribute 'X509_V_FLAG_CB_ISSUER_CHECK'

  ```
</details>

Looks from [stackoverflow](https://stackoverflow.com/questions/73830524/attributeerror-module-lib-has-no-attribute-x509-v-flag-cb-issuer-check) it might be an issue with pyopenssl/urllib3 (imported by boto3). This seems to be a [know issue](https://github.com/boto/boto3/issues/3585) in boto3. 

Could you please fix the requirements of the OSCAR python-client so that this error is fixed?

I leave here my current configuration of packages (including pyopenssl/urllib3) in case it helps:
<details>
  <summary>pip list</summary>
 anaconda-navigator                 2.0.3
anaconda-project                   0.9.1
analytics-python                   1.4.post1
ansi2html                          1.7.0
anyio                              3.6.1
apeye                              1.2.0
apispec                            3.3.2
appdirs                            1.4.4
argh                               0.26.2
argon2-cffi                        20.1.0
arrow                              1.2.2
asn1crypto                         1.4.0
astroid                            2.5
astropy                            4.2.1
astunparse                         1.6.3
async-generator                    1.10
async-timeout                      4.0.2
atomicwrites                       1.4.0
attrs                              20.3.0
autodocsumm                        0.2.8
autopage                           0.5.1
autopep8                           1.5.6
Babel                              2.9.0
backcall                           0.2.0
backoff                            1.10.0
backports.functools-lru-cache      1.6.4
backports.shutil-get-terminal-size 1.0.0
backports.tempfile                 1.0
backports.weakref                  1.0.post1
bandit                             1.7.4
bcrypt                             4.0.1
beautifulsoup4                     4.9.3
binaryornot                        0.4.4
bitarray                           2.1.0
bkcharts                           0.2
black                              23.3.0
bleach                             3.3.0
blinker                            1.5
bokeh                              2.3.2
boto                               2.49.0
boto3                              1.28.80
botocore                           1.31.80
Bottleneck                         1.3.2
branca                             0.5.0
Brotli                             1.0.9
brotlipy                           0.7.0
CacheControl                       0.12.11
cachetools                         5.3.1
cchardet                           2.1.7
certifi                            2020.12.5
cffi                               1.14.5
cftime                             1.5.2
chardet                            4.0.0
charset-normalizer                 2.0.12
clang                              5.0
click                              8.1.3
cliff                              3.10.1
cloudpickle                        1.6.0
clyent                             1.2.2
cmaes                              0.8.2
cmd2                               2.4.1
colorama                           0.4.6
colorlog                           6.6.0
commonmark                         0.9.1
conda                              4.10.1
conda-build                        3.21.4
conda-content-trust                0+unknown
conda-package-handling             1.7.3
conda-repo-cli                     1.0.4
conda-token                        0.3.0
conda-verify                       3.4.2
ConfigArgParse                     1.5.3
contextlib2                        0.6.0.post1
cookiecutter                       2.1.1
coverage                           6.4.1
cryptography                       40.0.1
css-html-js-minify                 2.5.5
cssutils                           2.4.2
cycler                             0.10.0
Cython                             0.29.23
cytoolz                            0.11.0
dash                               2.1.0
dash-bootstrap-components          0.13.1
dash-core-components               2.0.0
dash-html-components               2.0.0
dash-table                         5.0.0
dask                               2021.4.0
debtcollector                      2.5.0
decorator                          5.0.6
deep-dashboard                     0.9.7.dev2          /home/iheredia/ignacio/projects/deephdc/deep-dashboard
deepaas                            2.0.0               /home/iheredia/ignacio/projects/deephdc/deepaas
defusedxml                         0.7.1
demo-app                           0.0.1.dev22         /home/iheredia/ignacio/projects/deephdc/apps/demo_app
deprecation                        2.1.0
dict2css                           0.3.0
diff-match-patch                   20200713
distlib                            0.3.6
distributed                        2021.4.1
dm-reverb                          0.7.1
dm-tree                            0.1.7
docopt                             0.6.2
docutils                           0.17.1
domdf-python-tools                 3.3.0
entrypoints                        0.3
et-xmlfile                         1.0.1
extras                             1.0.0
fastapi                            0.89.1
fastcache                          1.1.0
fasteners                          0.18
fedserver                          0.0.1.dev16         /home/iheredia/ignacio/projects/deephdc/apps/federated-server
feedgenerator                      2.0.0
ffmpy                              0.3.0
filelock                           3.12.2
fixtures                           4.0.0
flaat                              1.1.8
flake8                             3.9.0
Flask                              2.2.3
Flask-CacheBuster                  1.0.0
Flask-Compress                     1.10.1
Flask-Cors                         3.0.10
Flask-Login                        0.6.2
flatbuffers                        1.12
flwr                               1.5.0
folium                             0.13.0
frozenlist                         1.3.0
fsspec                             0.9.0
future                             0.18.2
gast                               0.4.0
gevent                             21.1.2
ghp-import                         2.1.0
gin-config                         0.5.0
gitdb                              4.0.9
GitPython                          3.1.31
glob2                              0.7
gmpy2                              2.0.8
google-auth                        1.35.0
google-auth-oauthlib               0.4.6
google-pasta                       0.2.0
gradio                             2.3.9
greenlet                           1.0.0
grpcio                             1.56.0
gunicorn                           20.1.0
gym                                0.21.0
gym-notices                        0.0.6
h11                                0.12.0
h5py                               3.1.0
HeapDict                           1.0.1
html5lib                           1.1
httpcore                           0.15.0
httptools                          0.5.0
httpx                              0.23.0
huggingface-hub                    0.13.3
humanfriendly                      10.0
idna                               2.10
imageio                            2.9.0
imagesize                          1.2.0
imgaug                             0.4.0
importlib-metadata                 4.11.3
importlib-resources                5.7.1
iniconfig                          1.1.1
intervaltree                       3.1.0
ipykernel                          5.3.4
ipython                            7.22.0
ipython-genutils                   0.2.0
ipywidgets                         7.6.3
iso8601                            1.0.2
isort                              5.8.0
iterators                          0.0.2
itsdangerous                       2.1.2
jaal                               0.1.2
jdcal                              1.4.1
jedi                               0.17.2
jeepney                            0.6.0
Jinja2                             3.1.2
jinja2-time                        0.2.0
jmespath                           1.0.1
joblib                             1.0.1
json5                              0.9.5
jsonpickle                         2.2.0
jsonschema                         3.2.0
jupyter                            1.0.0
jupyter-client                     6.1.12
jupyter-console                    6.4.0
jupyter-core                       4.7.1
jupyter-dash                       0.4.0
jupyter-packaging                  0.7.12
jupyter-server                     1.4.1
jupyterlab                         3.0.14
jupyterlab-pygments                0.1.2
jupyterlab-server                  2.4.0
jupyterlab-widgets                 1.0.0
keras                              2.8.0
Keras-Preprocessing                1.1.2
keyring                            22.3.0
kiwisolver                         1.3.1
lazy-object-proxy                  1.6.0
libarchive-c                       2.9
libclang                           13.0.0
liboidcagent                       0.5.2
line-profiler                      3.5.1
line-profiler-pycharm              1.1.0
linkify-it-py                      2.0.0
llvmlite                           0.36.0
locket                             0.2.1
lockfile                           0.12.2
lxml                               4.9.2
Mako                               1.2.0
Markdown                           3.3.4
markdown-it-py                     2.2.0
markdown2                          2.4.8
MarkupSafe                         2.1.1
marshmallow                        3.16.0
matplotlib                         3.3.4
mccabe                             0.6.1
mdit-py-plugins                    0.3.3
mdurl                              0.1.2
mistune                            0.8.4
mkl-fft                            1.3.0
mkl-random                         1.2.1
mkl-service                        2.3.0
mock                               4.0.3
monotonic                          1.6
more-itertools                     8.7.0
mpmath                             1.2.1
msgpack                            1.0.2
multidict                          6.0.2
multipledispatch                   0.6.0
mutagen                            1.46.0
mypy-extensions                    0.4.3
natsort                            8.1.0
navigator-updater                  0.2.1
nbclassic                          0.2.6
nbclient                           0.5.3
nbconvert                          6.0.7
nbformat                           5.1.3
nest-asyncio                       1.5.1
netaddr                            0.8.0
netCDF4                            1.5.8
netifaces                          0.11.0
networkx                           2.5
nltk                               3.6.1
nose                               1.3.7
notebook                           6.3.0
numba                              0.53.1
numexpr                            2.7.3
numpy                              1.22.3
numpydoc                           1.1.0
oauthlib                           3.1.1
olefile                            0.46
opencv-python                      4.7.0.68
opencv-python-headless             4.7.0.72
openpyxl                           3.0.7
opt-einsum                         3.3.0
optuna                             2.10.0
orjson                             3.8.8
orpy                               1.0.2
oscar-python                       1.2.1
oslo.concurrency                   5.0.1
oslo.config                        8.8.0
oslo.context                       4.1.0
oslo.i18n                          5.1.0
oslo.log                           5.0.0
oslo.serialization                 4.3.0
oslo.utils                         4.13.0
packaging                          23.1
pandas                             1.4.1
pandocfilters                      1.4.3
paramiko                           3.1.0
parso                              0.7.0
partd                              1.2.0
path                               15.1.2
pathlib2                           2.3.5
pathspec                           0.11.1
patsy                              0.5.1
pbr                                5.9.0
pelican                            4.8.0
pep8                               1.7.1
pexpect                            4.8.0
pickleshare                        0.7.5
pikepdf                            7.1.1
Pillow                             9.4.0
pip                                23.3.1
pipreqs                            0.4.11
pkginfo                            1.7.0
platformdirs                       3.8.0
plotly                             5.6.0
pluggy                             0.13.1
ply                                3.11
portpicker                         1.5.0
prettytable                        3.3.0
prometheus-client                  0.10.1
prompt-toolkit                     3.0.17
protobuf                           3.20.3
psutil                             5.8.0
ptyprocess                         0.7.0
py                                 1.10.0
pyasn1                             0.4.8
pyasn1-modules                     0.2.8
pycares                            4.2.2
pycodestyle                        2.7.0
pycosat                            0.6.3
pycparser                          2.20
pycryptodome                       3.17
pycryptodomex                      3.15.0
pycurl                             7.43.0.6
pydantic                           1.10.4
pydocstyle                         6.0.0
pydub                              0.25.1
pyerfa                             1.7.3
pyflakes                           2.3.1
Pygments                           2.14.0
pygod                              0.3.0
pyinotify                          0.9.6
PyJWT                              2.6.0
pylint                             2.7.4
pyls-black                         0.4.6
pyls-spyder                        0.3.2
PyNaCl                             1.5.0
pyodbc                             4.0.0-unsupported
pyOpenSSL                          20.0.1
pyparsing                          2.4.7
pypdf                              3.5.1
pyperclip                          1.8.2
pyproj                             3.4.0
pyproject-api                      1.5.2
pyrsistent                         0.17.3
PySocks                            1.7.1
pyte                               0.8.1
pytest                             6.2.3
pytest-cov                         3.0.0
python-dateutil                    2.8.1
python-dotenv                      0.21.1
python-jsonrpc-server              0.4.0
python-language-server             0.36.2
python-Levenshtein                 0.12.2
python-multipart                   0.0.6
python-nomad                       2.0.0
python-slugify                     6.1.2
python-subunit                     1.4.0
pytz                               2021.1
pyvis                              0.2.1
PyWavelets                         1.1.1
pyxdg                              0.27
PyYAML                             6.0.1
pyzmq                              20.0.0
QDarkStyle                         2.8.1
QtAwesome                          1.0.2
qtconsole                          5.0.3
QtPy                               1.9.0
qudida                             0.0.4
recommonmark                       0.7.1
regex                              2021.4.4
requests                           2.25.1
requests-oauthlib                  1.3.0
retrying                           1.3.3
rfc3986                            1.5.0
rfc3987                            1.3.8
rich                               13.5.2
rope                               0.18.0
rsa                                4.7.2
Rtree                              0.9.7
ruamel.yaml                        0.17.21
ruamel.yaml.clib                   0.2.6
ruamel-yaml-conda                  0.15.100
ruff                               0.0.265
s3transfer                         0.7.0
schedule                           1.1.0
schema4deep                        0.0.1
scikit-image                       0.18.1
scikit-learn                       0.24.1
scipy                              1.8.0
seaborn                            0.11.1
SecretStorage                      3.3.1
semantic-version                   2.10.0
Send2Trash                         1.5.0
setuptools                         52.0.0.post20210125
shap                               0.40.0
shapely                            2.0.0
shtab                              1.5.8
simplegeneric                      0.8.1
simplejson                         3.17.6
singledispatch                     0.0.0
sip                                4.19.13
six                                1.15.0
slicer                             0.0.7
smmap                              5.0.0
sniffio                            1.2.0
snowballstemmer                    2.1.0
sortedcollections                  2.1.0
sortedcontainers                   2.3.0
soupsieve                          2.2.1
Sphinx                             4.5.0
sphinx-autodoc-typehints           1.18.3
sphinx_design                      0.4.1
sphinx-jinja2-compat               0.1.2
sphinx-markdown-tables             0.0.15
sphinx-material                    0.0.35
sphinx-prompt                      1.5.0
sphinx-rtd-theme                   1.0.0
sphinx-tabs                        3.3.1
sphinx-toolbox                     3.1.1
sphinxcontrib-applehelp            1.0.2
sphinxcontrib-devhelp              1.0.2
sphinxcontrib-htmlhelp             2.0.0
sphinxcontrib-jsmath               1.0.1
sphinxcontrib-qthelp               1.0.3
sphinxcontrib-serializinghtml      1.1.5
sphinxcontrib-websupport           1.2.4
spyder                             4.2.5
spyder-kernels                     1.10.2
SQLAlchemy                         1.4.15
stable-baselines3                  1.5.0
starlette                          0.22.0
statsmodels                        0.12.2
stestr                             3.2.1
stevedore                          3.5.0
sympy                              1.8
tables                             3.6.1
tabulate                           0.8.9
tblib                              1.7.0
tenacity                           8.0.1
tensorboard                        2.8.0
tensorboard-data-server            0.6.1
tensorboard-plugin-wit             1.8.0
tensorboardX                       2.5.1
tensorflow                         2.8.0
tensorflow-estimator               2.6.0
tensorflow-io-gcs-filesystem       0.24.0
tensorflow-probability             0.16.0
termcolor                          1.1.0
terminado                          0.9.4
testpath                           0.4.4
testtools                          2.5.0
text-unidecode                     1.3
textdistance                       4.2.1
tf-agents                          0.12.0
tf-estimator-nightly               2.8.0.dev2021122109
thefuck                            3.32
thefuzz                            0.19.0
threadpoolctl                      2.1.0
three-merge                        0.1.1
tianshou                           0.4.7
tifffile                           2020.10.1
tldr                               3.1.0
toml                               0.10.2
tomli                              2.0.1
toolz                              0.11.1
torch                              1.11.0
torch-cluster                      1.6.0
torch-geometric                    2.1.0.post1
torch-scatter                      2.0.9
torch-sparse                       0.6.15
torch-spline-conv                  1.2.1
torchaudio                         0.11.0
torchvision                        0.12.0
tornado                            6.1
tox                                3.9.0
tqdm                               4.59.0
traitlets                          5.0.5
typed-ast                          1.4.2
typer                              0.7.0
typing_extensions                  4.4.0
uc-micro-py                        1.0.1
ujson                              4.0.2
unicodecsv                         0.14.1
Unidecode                          1.3.6
urllib3                            1.26.4
uvicorn                            0.20.0
uvloop                             0.17.0
virtualenv                         20.23.1
visdcc                             0.0.50
voluptuous                         0.13.1
watchdog                           1.0.2
watchfiles                         0.18.1
waybackpy                          3.0.6
wcwidth                            0.2.5
webargs                            5.5.3
webdavclient3                      3.14.6
webencodings                       0.5.1
websockets                         10.4
Werkzeug                           2.2.3
wheel                              0.36.2
widgetsnbextension                 3.5.1
wrapt                              1.12.1
wurlitzer                          2.1.0
xlrd                               2.0.1
XlsxWriter                         1.3.8
xlwt                               1.3.0
xmltodict                          0.12.0
yapf                               0.31.0
yarg                               0.1.9
yarl                               1.7.2
yt-dlp                             2022.10.4
zict                               2.0.0
zipp                               3.4.1
zope.event                         4.5.0
zope.interface                     5.3.0

  ```
</details>

**Comment by IgnacioHeredia** [comment]
In addition, a more general comment. 

When users interact with services they should be able to provide a VO as the parameter of the function. Because a single user can belong to several VOs (ai4eosc, imagine-ai), so when you create the service you should be able to specify to which VO it is assigned. This is also important for accounting because you might want to consume the resources assigned to one VO and not another.
Same with retrieving services, you should retrieve only the services deployed for a given VO (optionally giving the option to retrieve services of all VOs).

You can use the [deployments file](https://github.com/ai4os/ai4-papi/blob/master/ai4papi/routers/v1/deployments/modules.py) as a reference on how this was implemented.

Not familiar with OSCAR but I see three options:
1. keep using one OSCAR cluster for all VOs, and have namespaces for each VO (are namespaces possible in OSCAR?)
2. keep using one OSCAR cluster, and add the VO to the service metadata. The con of this approach is that in order to retrieve the services belonging to one VO, you have to retrieve _all_ the services then filter based on the metadata.
3. use one OSCAR cluster per VO

Maybe @gmolto can comment on this.

**Comment by IgnacioHeredia** [comment]
Great! Let me know we it is done.

**Comment by micafer** [comment]
We have tested it in a clean installation on a docker container (ubuntu 22.04), and it works, as pyOpenSSL is not installed, because pyOpenSSL is not a requirement of OSCAR or boto3.
The problem seems to arise when the current environment has an old version of pyOpenSSL already installed.
pyOpenSSL is not a requirement of OSCAR. It is a problem of urllib3 and pyOpenSSL libs. So we should not add a dependency on OSCAR (that we do not use) in our tool just to solve a problem of external libraries.

**Comment by IgnacioHeredia** [comment]
I see. I can confirm that upgrading pyOpenSSL fixes the issue, so problem closed. Thanks @micafer !

**Comment by IgnacioHeredia** [comment]
Hi @catttam , let me know when this is ready for review!

**Comment by catttam** [comment]
> Hi @catttam , let me know when this is ready for review!

Hi, it is ready for review :)

**Comment by IgnacioHeredia** [comment]
Thanks @catttam !
